File: pet-backend/src/modules/auth/auth.service.ts
import { Injectable, UnauthorizedException, ConflictException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import * as bcrypt from 'bcrypt';

import { UsersService } from '../users/users.service';

@Injectable()
export class AuthService {
  constructor(
    private readonly usersService: UsersService,
    private readonly jwtService: JwtService,
  ) {}

  async register(dto: { name: string; email: string; password: string }) {
    const existing = await this.usersService.findByEmail(dto.email);
    if (existing) {
      throw new ConflictException('Email already registered');
    }

    const passwordHash = await bcrypt.hash(dto.password, 12);
    const user = await this.usersService.create({
      name: dto.name,
      email: dto.email,
      passwordHash,
    });

    return {
      user: { id: user.id, name: user.name, email: user.email },
      token: this.generateToken(user.id, user.email),
    };
  }

  async login(dto: { email: string; password: string }) {
    const user = await this.usersService.findByEmail(dto.email);
    if (!user) {
      throw new UnauthorizedException('Invalid credentials');
    }

    const isPasswordValid = await bcrypt.compare(dto.password, user.passwordHash);
    if (!isPasswordValid) {
      throw new UnauthorizedException('Invalid credentials');
    }

    return {
      user: { id: user.id, name: user.name, email: user.email, role: user.role },
      token: this.generateToken(user.id, user.email),
    };
  }

  private generateToken(userId: string, email: string): string {
    return this.jwtService.sign({ sub: userId, email });
  }
}

File: pet-backend-fastapi/app/services/auth_service.py
from datetime import datetime, timezone

from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from sqlalchemy.orm import selectinload

from app.models.user import User
from app.utils.security import hash_password, verify_password, create_access_token
from app.schemas.auth import RegisterRequest, LoginRequest


class AuthService:
    def __init__(self, db: AsyncSession):
        self.db = db

    async def register(self, dto: RegisterRequest) -> dict:
        existing = await self.db.execute(select(User).where(User.email == dto.email))
        if existing.scalar_one_or_none():
            raise ValueError("Email already registered")

        user = User(
            email=dto.email,
            password_hash=hash_password(dto.password),
            name=dto.name,
        )
        self.db.add(user)
        await self.db.commit()
        await self.db.refresh(user)

        return {
            "token": create_access_token({"sub": str(user.id), "email": user.email}),
            "user": user,
        }

    async def login(self, dto: LoginRequest) -> dict:
        result = await self.db.execute(select(User).where(User.email == dto.email))
        user = result.scalar_one_or_none()

        if not user or not verify_password(dto.password, user.password_hash):
            raise ValueError("Invalid credentials")

        return {
            "token": create_access_token({"sub": str(user.id), "email": user.email}),
            "user": user,
        }
